Rewrite MEP doc as additional, not either/or (#1665)
globus/globus-compute: docs/endpoints/endpoints.rst

Globus Compute Endpoints
************************

A Globus Compute Endpoint is an agent (i.e., a service; a process) launched by the user
to serve as a conduit for executing functions on a computing resource.  The Compute

> Endpoint agent (CEA) manages the site‑specific interactions for executing functions,

Author: > I think we should reserve the term "agent" when specifically referring to the installable package. In most cases, including this one, we refer to the process running on the machine, which we can simply call the Compute Endpoint.

That ... took more wordsmithing than I had thought it would.

> I also recommend avoiding acronyms like CEA. A user that, for example, jumps directly to the Getting Started section will not know what CEA stands for.

Yeah, I've been waffling on this as well.  I'm not so concerned about jumping around -- just search for it and learn it -- but the larger point for me is that CEA is just awkward as an acronym.  Moreover, we're not strictly in the academic realm so probably safe to just elide it completely.

> Endpoint agent (CEA) manages the site‑specific interactions for executing functions,

Reviewer: Yea a new acronym would be confusing if it's not widely known, unless its definition is above uses on the same page, to reduce lengthy redundancy.  I don't think it was repeated too many times here?  So 👍 to hide it (except for internal links that you are using it for)

the service:

.. code-block:: console

   $ globus-compute-endpoint disable-on-boot my-endpoint
   Run the following to disable on-boot-persistence:
      systemctl stop globus-compute-endpoint-my-endpoint
      systemctl disable globus-compute-endpoint-my-endpoint
      rm /etc/systemd/system/globus-compute-endpoint-my-endpoint.service

AMQP Port
---------

Endpoints receive tasks and communicate task results via the AMQP messaging protocol.
As of v2.11.0, newly configured endpoints use AMQP over port 443 by default, since
firewall rules usually leave that port open. In case 443 is not open on a particular
cluster, the port to use can be changed in the endpoint config via the ``amqp_port``
option, like so:

.. code-block:: yaml

   amqp_port: 5671
   display_name: My Endpoint
   engine: ...

Note that only ports 5671, 5672, and 443 are supported with the Compute hosted services.
Also note that when ``amqp_port`` is omitted from the config, the port is based on the
connection URL the endpoint receives after registering itself with the services, which
typically means port 5671.

.. _endpoints_templating_configuration:

Templating Endpoint Configuration

> ---------------------------------

Author: I thought about that, but these are two different use-cases.  For the most part, the multi-user endpoint documentation is oriented toward administrators, and I don't want to point "regular users" there.  In particular, note the very top of the multi-user section, which has this note:

<img src="https://github.com/user-attachments/assets/5b9aab76-c57d-446d-95ed-96e9a699a0b5" width="500" title="The Multi-User Endpoint page is oriented toward administrators, not users" alt="screenshot of top of multi-user-endpoint documentation page; showing the page path, title, and called-out note">

Globus Compute Endpoints
************************

A Globus Compute Endpoint is a process launched by the user to serve as a conduit for
executing functions on a computing resource.  The Compute Endpoint process manages the
site‑specific interactions for executing functions, leaving users with only a single API
necessary for running code on their laptop, campus cluster, or even leadership‑class HPC
machines.

> In the mental model of Globus Compute, Endpoints are the remote instance:

Reviewer: ```suggestion
In the mental model of Globus Compute, Endpoints runs on the user-accessible remote instance:
```

to more clearly delineate between Compute and user-land?

Author: I ... don't think so, not here.  This section is about framing the mental model.  The bullet point below delineates the user-accessible part.

  authenticate and ferry functions and tasks

- **Globus Compute Endpoint** |nbsp| --- |nbsp| a site-specific process, typically on a
  cluster head node, that manages user‑accessible compute resources to run tasks
  submitted by the SDK

The Compute Endpoint may be installed via PyPI in the usual manner, as well as via
system packages.  See :doc:`installation` for more information.

Quickstart
==========

For those just looking for the quickstart commands:

.. code-block:: console

   $ python3 -m pipx install globus-compute-endpoint

   $ globus-compute-endpoint configure my_first_endpoint

   $ globus-compute-endpoint start my_first_endpoint

   $ globus-compute-endpoint stop my_first_endpoint

Getting Started
===============

Creating New Compute Endpoints
------------------------------

Create a new endpoint directory and default files in ``$HOME/.globus_compute/`` via the
``configure`` subcommand:

.. code-block:: console

   $ globus-compute-endpoint configure my_first_endpoint
   Created profile for endpoint named <my_first_endpoint>

       Configuration file: /.../.globus_compute/my_first_endpoint/config.yaml

> The Compute Endpoint will also be in the output of the ``list`` subcommand:

Reviewer: ```suggestion
This new Compute Endpoint will also be in the output of the ``list`` subcommand:
```

.. code-block:: console

   $ globus-compute-endpoint list
   +--------------------------------------+--------------+--------------------+
   |             Endpoint ID              |    Status    |   Endpoint Name    |
   +======================================+==============+====================+
   | None                                 | Initialized  | my_first_endpoint  |
   +--------------------------------------+--------------+--------------------+

The Compute endpoint will receive an endpoint identifier from the Globus Compute web
service upon first connect.  Until then, it exists solely as a subdirectory of
``$HOME/.globus_compute/``.

.. _cea_configuration:

As Globus Compute endpoints may be run on a diverse set of computational resources
(i.e., the gamut from laptops to supercomputers), it is important to configure each
instance to match the underlying capabilities and restrictions of the resource.  The

> default configuration is functional |nbsp| --- |nbsp| it will process tasks |nbsp| ---

Reviewer: ```suggestion
default configuration below is functional |nbsp| --- |nbsp| it will process tasks |nbsp| ---
```

Author: I think "below" in this context would slightly open the room for "is one default different than another default?"  In my mind, there is only one default configuration, and I want to impart that here.  _The_ default configuration ...

> |nbsp| but it is intentionally limited to only use processes on the Endpoint host.  In

Reviewer: ```suggestion
|nbsp| but it is intentionally limited to only use processes on the Endpoint process host, not on additional ``computational`` nodes on a HPC.  In
```

Author: Wordsmithed this slighty differently.  Will push shortly.

its entirety, the default configuration is:

.. code-block:: yaml
   :caption: ``$HOME/.globus_compute/my_first_endpoint/config.yaml``

   display_name: null
   engine:
     max_workers_per_node: 1
     type: GlobusComputeEngine
     provider:
       type: LocalProvider
       init_blocks: 1
       max_blocks: 1
       min_blocks: 0

For now, the key items to observe are the structure (e.g., ``provider`` as a child of
``engine``), the engine type, ``GlobusComputeEngine``, and the provider type,
``LocalProvider``.

The *engine* pulls tasks from the incoming queue and conveys them to the *provider* for
execution.  Globus Compute implements three engines: ``ThreadPoolEngine``,
``ProcessPoolEngine``, and ``GlobusComputeEngine``.  The first two are Compute endpoint
wrappers of Python's |ThreadPoolExecutor|_ and |ProcessPoolExecutor|_.  These engines are

> most appropriate for single‑host (e.g., a personal workstation) installations.  For

Reviewer: ```suggestion
most appropriate for single‑host installations (e.g., a personal workstation).  For
```

.. code-block:: yaml

   amqp_port: 443
   display_name: Retry_2_times
   engine:
       type: GlobusComputeEngine
       max_retries_on_system_failure: 2  # Default=0

Auto-Scaling
^^^^^^^^^^^^

|GlobusComputeEngine|_ by default automatically scales workers in response to workload.

``Strategy`` configuration is limited to two options:

#. ``max_idletime``: Maximum duration in seconds that workers are allowed to idle before
   they are marked for termination

#. ``strategy_period``: Set the # of seconds between strategy attempting auto-scaling
   events

The bounds for scaling are determined by the options to the ``Provider``
(``init_blocks``, ``min_blocks``, ``max_blocks``). Please refer to the `Parsl docs
<https://parsl.readthedocs.io/en/stable/userguide/execution.html#elasticity>`_ for more
info.

Here's an example configuration:

.. code-block:: yaml

   engine:
       type: GlobusComputeEngine
       job_status_kwargs:
           max_idletime: 60.0      # Default = 120s
           strategy_period: 120.0  # Default = 5s

Ensuring Execution Environment
------------------------------

When executing a function, endpoint *worker processes* expect to have all dependencies

> readily available.  For example, if a function requires ``numpy`` and a worker

Reviewer 2: s/readily available/installed ?

As a result, it is often necessary to load in some kind of pre‑initialized environment
for each worker.  In general there are two approaches:

.. note::

   The worker environment must have the ``globus-compute-endpoint`` Python module
   installed.  We recommend matching the Python version and ``globus-compute-endpoint``
   module version on the worker environment and on the endpoint interchange.

1. Python-Based Environment Isolation
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Python‑based environment management uses the |worker_init|_ config option:

.. code-block:: yaml

   engine:
     provider:
       worker_init: |
         conda activate my-conda-env  # or venv, or virtualenv, or ...
         source /some/other/config

Though the exact behavior of ``worker_init`` depends on the specific |Provider|_, this
is typically run in the same process as (or the parent process of) the worker, allowing
environment modification (i.e., environment variables).

In some cases, it may also be helpful to run some setup during the startup process of
the endpoint itself, before any workers start.  This can be achieved using the top‑level
``endpoint_setup`` config option:

.. code-block:: yaml

>    endpoint_setup: |

Reviewer 2: Could we make this config also show working_init being configured to activate the env for completeness?

Author: We do; look just above in the rendered documentation:

<img src="https://github.com/user-attachments/assets/4d712f48-66e7-4125-a43e-79e3eafd1006" title="" alt="" width="500">

Reviewer 2: This looks obvious in the rendered form. No changes necessary.

     conda create -n my-conda-env
     conda activate my-conda-env
     pip install -r requirements.txt

.. note::

   The script specified by ``endpoint_setup`` runs in a shell (usually ``/bin/sh``), as
   a child process, and must finish successfully before the endpoint will continue

>    starting up.  In particular, *note that it is not possible to use this hook to set or

Reviewer 2: I'd make this a warning.

   amqp_port: 5671
   display_name: My Endpoint
   engine: ...

Note that only ports 5671, 5672, and 443 are supported with the Compute hosted services.
Also note that when ``amqp_port`` is omitted from the config, the port is based on the
connection URL the endpoint receives after registering itself with the services, which
typically means port 5671.

.. _endpoints_templating_configuration:

Templating Endpoint Configuration
---------------------------------

A common experience for Compute users is a proliferation of their Compute Endpoints.
After starting with a basic configuration, changing or conflicting requirements
necessitate running multiple endpoints |nbsp| --- |nbsp| sometimes simultaneously |nbsp|
--- |nbsp| with different configurations.  For example, attributing work to different
accounts, changing the size of the provisioned compute resource to match the demands of
the problem, or changing the submission queue to the batch‑system.

This becomes a bit of an administrative mess for these users, who must constantly update
their endpoint configurations, bring up and bring down different endpoints, and be aware
of which endpoints have which configuration.

> As of May, 2024, Compute Endpoints may now be run as "multi‑user" endpoints.  Please

Reviewer: I looked for 10s and couldn't find the 'note' below that's obviously referred to (assuming it's the immediately following paragraph), so I'd say the words aren't needed? ie.

```suggestion
As of May, 2024, Compute Endpoints may now be run as "multi‑user" endpoints.  Ignore the 'multi' part of the name, and think of it as a "template-able" user endpoint.
```

Author: I'll link to the note instead then.  It's at the end of the section.

> ignore the name (see the note, below) and instead think of it as "template‑able".  A

Reviewer: ```suggestion
ignore the name (see the note, below) and instead think of it as "template‑able".  This type of 'multi'-user endpoint
```

multi‑user endpoint specifies a configuration *template* that will be filled in by

> SDK‑supplied user‑variables.  This configuration is then applied to sub‑processes of the

Reviewer: ```suggestion
SDK‑supplied user‑variables.  This configuration is then applied to sub‑processes of the endpoint.
```

Author: I think this is already ... correct as is?

Create a MEP configuration by passing the ``--multi-user`` flag to the ``configure``
subcommand:

.. code-block:: console

   $ globus-compute-endpoint configure --multi-user my_second_endpoint
   Created multi-user profile for endpoint named <my_second_endpoint>

       Configuration file: /.../.globus_compute/my_second_endpoint/config.yaml

       Example identity mapping configuration: /.../.globus_compute/my_second_endpoint/example_identity_mapping_config.json

       User endpoint configuration template: /.../.globus_compute/my_second_endpoint/user_config_template.yaml.j2
       User endpoint configuration schema: /.../.globus_compute/my_second_endpoint/user_config_schema.json
       User endpoint environment variables: /.../.globus_compute/my_second_endpoint/user_environment.yaml

     Use the `start` subcommand to run it:

       $ globus-compute-endpoint start my_second_endpoint

The default configuration of the MEP, in its entirety, is:

.. code-block:: console

   $ cat /.../.globus_compute/my_second_endpoint/config.yaml
   display_name: null
   identity_mapping_config_path: /.../.globus_compute/my_second_endpoint/example_identity_mapping_config.json
   multi_user: true

> Unless this MEP will be run as a power user (e.g., ``root``) |nbsp| --- |nbsp| in which

Reviewer: ```suggestion
Unless this MEP will be run as a privileged user (e.g., ``root``) |nbsp| --- |nbsp| in which
```

> case, please read :doc:`the next section <multi_user>` |nbsp| --- |nbsp| the Identity

Reviewer: ```suggestion
case, please read :doc:`the next section <multi_user>` |nbsp| --- |nbsp|, the Identity
```

Author: No, not after the mdash.
